packages/gateway/src/durable-objects/metrics.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @typedef {Object} GatewayMetrics
+ * @property {number} totalReqCount total number of performed requests
+ * @property {number} failedReqCount total number of requests failed
+ * @property {number} fasterReqCount number of performed requests where faster
+ * @property {Record<string, number>} responseTimeHistogram
+ * @property {number} [averageResponseTime]
+ *
+ * @typedef {Object} ResponseStats
+ * @property {string} url gateway URL
+ * @property {boolean} ok request was successful
+ * @property {number} [responseTime] number of milliseconds to get response
+ * @property {boolean} [faster]
+ */
+
+/**
+ * Durable Object for keeping Metrics state.
+ */
+export class Metrics8 {
+  constructor(state, env) {
+    this.state = state
+    /** @type {Array<string>} */
+    this.ipfsGateways = JSON.parse(env.IPFS_GATEWAYS)
+
+    // `blockConcurrencyWhile()` ensures no requests are delivered until initialization completes.
+    this.state.blockConcurrencyWhile(async () => {
+      /** @type {Map<string, GatewayMetrics>} */
+      this.gatewayMetrics = new Map()
+
+      // Get state and initialize if not existing
+      const storedMetricsPerGw = await Promise.all(
+        this.ipfsGateways.map(async (gw) => {
+          /** @type {GatewayMetrics} */
+          const value = (await this.state.storage.get(gw)) || {
+            ...defaultCounter,
```

Comment:
Needs to clone `responseTimeHistogram` from `defaultCounter` or all of them share the same one. Probably best to make this into a class and do `new Counter` instead to sidestep any cloning issues.

Reply:
Somehow I think they are not sharing the same reference https://ipfs-staging.nft.storage/metrics but you are right, it should be and we need a better solution.

The thing is, Durable objects only accept POJO and not class instances to be stored. So, I think the best solution is to have a create function such as:

```js
const histogram = [300, 500, 750, 1000, 1500, 2000, 3000, 5000, 10000]

function createMetricsTracker () {
  const h = [...histogram].map((h) => [h, 0])

  return {
    totalReqCount: 0,
    failedReqCount: 0,
    fasterReqCount: 0,
    responseTimeHistogram: Object.fromEntries(h),
    averageResponseTime: undefined,
  }
}
```

Reply:
Yes, or that :)

You don't need to spread `histogram` though you can just `histogram.map`.